contracts/andromeda_vault/src/contract.rs

```diff
@@ -0,0 +1,426 @@
+                    BALANCES.save(
+                        deps.storage,
+                        (info.sender.to_string(), denom),
+                        &balance.checked_sub(amount)?,
+                    )?;
+                }
+            },
+            None => {
+                withdrawal_amount.push(coin(balance.u128(), denom.clone()));
+                BALANCES.save(
+                    deps.storage,
+                    (info.sender.to_string(), denom),
+                    &Uint128::zero(),
+                )?;
+            }
+        }
+    }
+    let withdrawal_msg = CosmosMsg::Bank(BankMsg::Send {
+        to_address: recipient,
+        amount: withdrawal_amount,
+    });
+    Ok(res.add_message(withdrawal_msg))
+}
+
+pub fn withdraw_strategy(
+    deps: DepsMut,
+    info: MessageInfo,
+    strategy: StrategyType,
+    recipient: Option<Recipient>,
+    withdrawals: Vec<Withdrawal>,
+) -> Result<Response, ContractError> {
+    let res = Response::default();
+    let recipient = recipient.unwrap_or_else(|| Recipient::Addr(info.sender.to_string()));
+    let addr_opt = STRATEGY_CONTRACT_ADDRESSES.may_load(deps.storage, strategy.to_string())?;
+    if addr_opt.is_none() {
+        return Err(ContractError::InvalidStrategy {
+            strategy: strategy.to_string(),
+        });
+    }
+
+    let addr = addr_opt.unwrap();
+    let withdraw_exec = to_binary(&ExecuteMsg::AndrReceive(AndromedaMsg::Withdraw {
+        recipient: Some(recipient),
+        tokens_to_withdraw: Some(withdrawals),
+    }))?;
+    let withdraw_submsg = SubMsg {
+        id: 104,
+        msg: CosmosMsg::Wasm(WasmMsg::Execute {
+            contract_addr: addr,
+            msg: withdraw_exec,
+            funds: vec![],
+        }),
+        gas_limit: None,
+        reply_on: ReplyOn::Error,
+    };
+
+    Ok(res.add_submessage(withdraw_submsg))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn migrate(deps: DepsMut, _env: Env, _msg: MigrateMsg) -> Result<Response, ContractError> {
+    let version = get_contract_version(deps.storage)?;
+    if version.contract != CONTRACT_NAME {
+        return Err(ContractError::CannotMigrate {
+            previous_contract: version.contract,
+        });
+    }
+    Ok(Response::default())
+}
+
+#[entry_point]
+pub fn query(deps: Deps, env: Env, msg: QueryMsg) -> Result<Binary, ContractError> {
+    match msg {
+        QueryMsg::AndrQuery(msg) => handle_andromeda_query(deps, env, msg),
+        QueryMsg::Balance {
+            address,
+            strategy,
+            denom,
+        } => query_balance(deps, env, address, strategy, denom),
+        QueryMsg::StrategyAddress { strategy } => query_strategy_address(deps, env, strategy),
+    }
+}
+
+fn handle_andromeda_query(
+    deps: Deps,
+    env: Env,
+    msg: AndromedaQuery,
+) -> Result<Binary, ContractError> {
+    ADOContract::default().query(deps, env, msg, query)
+}
```

Comment:
If we only use the default query I think we should inline the call.

Reply:
Added a default GET for balance by address